File: Heuristic Search Planner (sgware)/src/edu/uno/ai/planning/hsp/HSPHeuristic.java
package edu.uno.ai.planning.hsp;

import java.util.HashMap;

import edu.uno.ai.planning.State;
import edu.uno.ai.planning.Step;
import edu.uno.ai.planning.h.StateHeuristic;
import edu.uno.ai.planning.logic.Conjunction;
import edu.uno.ai.planning.logic.Expression;
import edu.uno.ai.planning.logic.Literal;
import edu.uno.ai.planning.ss.StateSpaceProblem;

public abstract class HSPHeuristic extends StateHeuristic {

	public static final int INFINITY = Integer.MAX_VALUE;
	private final HashMap<Literal, Integer> cost = new HashMap<>();
	
	public HSPHeuristic(StateSpaceProblem problem) {
		super(problem);
	}
	
	protected final int cost(Literal literal) {
		Integer i = cost.get(literal);
		if(i == null)
			return INFINITY;
		else
			return i;
	}
	
	protected abstract int cost(Expression expression);
	
	@Override
	public double evaluate(State current) {
		cost.clear();
		for(Step step : problem.steps)
			if(step.precondition.isTrue(current))
				setCost(step.precondition, 0);
		boolean again = true;
		while(again && cost(problem.goal) == INFINITY) {
			again = false;
			for(Step step : problem.steps) {
				int cost = cost(step.precondition);
				if(cost != INFINITY)
					again = setCost(step.effect, cost + 1) || again;
			}
		}
		return cost(problem.goal);
	}
	
	private final boolean setCost(Expression expression, int value) {
		if(expression instanceof Literal) {
			Literal literal = (Literal) expression;
			int current = cost(literal);
			if(value < current) {
				cost.put((Literal) expression, value);
				return true;
			}
			else
				return false;
		}
		else if(expression instanceof Conjunction) {
			boolean result = false;
			for(Expression argument : ((Conjunction) expression).arguments)
				result = setCost(argument, value) || result;
			return result;
		}
		else
			throw new UnsupportedOperationException(expression.getClass() + " not supported.");
	}
}

File: UnitySokoban/Assets/Scripts/Planning/HeuristicSearchPlannerSGW/HSPHeuristic.cs
using Planning;
using Planning.Logic;
using Planning.Util;
using StateSpaceSearchProject;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace HeuristicSearchPlannerSGW
{
    public abstract class HSPHeuristic : StateHeuristic
    {

        public static readonly int INFINITY = int.MaxValue;
        private readonly ImmutableArray<Literal> literals;
        private readonly Dictionary<Literal, int> costMap = new Dictionary<Literal, int>();

        public HSPHeuristic(StateSpaceProblem problem)
            : base(problem)
        {
            List<Literal> literals = new List<Literal>();
            foreach (Step step in problem.steps)
            {
                everyLiteral(step.precondition, (literal)=> { literals.Add(literal); });
                everyLiteral(step.effect, (literal)=> { literals.Add(literal); });
            }
            this.literals = new ImmutableArray<Literal>(literals.ToArray());
        }

        private static void everyLiteral(Expression expression, Action<Literal> consumer) {
            if (expression is Literal)
			consumer.Invoke((Literal)expression);
		else if (expression is Conjunction)
			foreach (Expression argument in ((Conjunction)expression).arguments)
                everyLiteral(argument, consumer);
		else
			throw new InvalidOperationException(expression.GetType() + " not supported.");
        }

        protected virtual int cost(Literal literal) {
            if (costMap.ContainsKey(literal))
                return costMap[literal];
            else
                return INFINITY;
        }

        protected abstract int cost(Expression expression);

        public override int evaluate(State current)
        {
            costMap.Clear();
            foreach (Literal literal in literals)
            {
                if (current.isTrue(literal))
                    costMap[literal] = 0;
            }
            List<Step> steps = new List<Step>();
            foreach (Step step in problem.steps)
                steps.Add(step);
            bool again = true;
            while (again && cost(problem.goal) == INFINITY)
            {
                again = false;
                IEnumerator<Step> iterator = steps.GetEnumerator();
                while (iterator.MoveNext())
                {
                    Step step = iterator.Current;
                    int costValue = cost(step.precondition);
                    if (costValue != INFINITY)
                    {
                        again = setCost(step.effect, costValue + 1) || again;
                        //iterator.remove();
                    }
                }
            }
            return cost(problem.goal);
        }

        private bool setCost(Expression expression, int value) {
            if (expression is Literal) {
                Literal literal = (Literal)expression;
                int current = cost(literal);
                if (value < current)
                {
                    costMap[literal] = value;
                    return true;
                }
                else
                    return false;
            }
		else if (expression is Conjunction) {
                bool result = false;
                foreach (Expression argument in ((Conjunction)expression).arguments)
                    result = setCost(argument, value) || result;
                return result;
            }
		else
			throw new InvalidOperationException(expression.GetType() + " not supported.");
        }
    }

}
